src/structured-text/structured_text_parser.cpp

```diff
@@ -0,0 +1,269 @@
+/*******************************************************************\
+
+Module: Structured Text Language Parser
+
+Author: Matthias Weiss, matthias.weiss@diffblue.com
+
+\*******************************************************************/
+
+/// \file
+/// Structured Text Language Parser
+
+#include "structured_text_parser.h"
+#include "structured_text_parse_tree.h"
+#include <cmath>
+#include <iostream>
+#include <util/expr.h>
+
+structured_text_parsert structured_text_parser;
+
+extern char *yystructured_texttext;
+
+/// Forwards any errors that are encountered during the parse process. This
+/// function gets called by the generated files of flex and bison.
+int yystructured_texterror(const std::string &error)
+{
+  structured_text_parser.parse_error(error, yystructured_texttext);
+  return 0;
+}
+
+void structured_text_parsert::add_function_block(exprt &block)
+{
+  INVARIANT(
+    block.id() == ID_structured_text_function_block,
+    "Expression ID should be structured_text_function_block");
+
+  current_root_element = block;
+
+  // Search for the correct tree elements and set the internal parser state.
+  INVARIANT(find_name(), "Function block should have a name");
+  INVARIANT(find_version(), "Function block should have a version");
+  INVARIANT(find_variable_list(), "Function block should have a variable list");
+  INVARIANT(find_network_list(), "Function block should have a network list");
+
+  // Generate new function block.
+  structured_text_parse_treet::function_blockt fb(
+    current_name, current_version);
+  current_parsed_root_element = &fb;
+
+  // Fill the block with networks and variables.
+  iterate_variables();
+  iterate_networks();
+
+  parse_tree.add_function_block(fb);
+}
+
+void structured_text_parsert::add_function(exprt &function)
+{
+  INVARIANT(
+    function.id() == ID_structured_text_function,
+    "Expression ID should be stl_function");
+
+  current_root_element = function;
+
+  // Search for the correct tree elements and set the internal parser state.
+  INVARIANT(find_name(), "Function should have a name");
```

Comment:
Since the whole function is created in the scope of the invocation of `add_function`, there is no need to work with members. Just have `find_version` etc. return their values, and move the invariant inside those functions.